Cancel pending request on de-registration

sbernard31 opened the pull request:
(This should fix the #77 )

boaks commented:
Would it be possible to put this change on an "experimental" branch?

boaks commented on leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java:
```diff
@@ -108,23 +117,102 @@ public T buildResponse(final Response coapResponse) {
+        if (requests != null) {
+            for (Request coapRequest : requests) {
+                coapRequest.cancel();
+                removePendingRequest(registrationId, coapRequest);
```
Did you check, if it's OK to remove something from a collection while "iterating" over it?

sbernard31 replied:
The test I write is ok, but this is clearly a mistake. (I will fix this)

boaks commented on leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java:
```diff
@@ -108,23 +117,102 @@ public T buildResponse(final Response coapResponse) {
+        ArrayList<Request> requests = pendingRequests.get(registrationId);
+        if (requests != null) {
+            for (Request coapRequest : requests) {
+                coapRequest.cancel();
```
Is it really intended to call "thirdparty" stuff within a "synchronized" section?
Even, if at least logicially, this "calls back" into the same synchronized section?

sbernard31 replied:
It should be fixed too.

boaks commented on leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java:
```diff
@@ -108,23 +117,102 @@ public T buildResponse(final Response coapResponse) {
         endpoint.sendRequest(coapRequest);
     }
 
+    public synchronized void cancelPendingRequests(String registrationId) {
```
Does it scale to introduce a "server global synchronized" section?

sbernard31 replied:
The map is shared between the server.
 What do you propose ?

boaks commented on leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java:
```diff
@@ -47,6 +51,7 @@
     private final Set<Endpoint> endpoints;
     private final ObservationRegistry observationRegistry;
     private final LwM2mModelProvider modelProvider;
+    private final Map<String/* registration id */, ArrayList<Request> /* pending coap Request */> pendingRequests = new HashMap<>();
```
I'm not sure, if an additional map with key <reg id> is really required.
Ok, it "isolates the concerns" but extending the "Client" may be a alternative, which is also not too bad.

sbernard31 replied:
Not sure to understand.  You want to extend the `org.eclipse.leshan.server.client.Client` ?
I'm not sure this is a good idea as Client should be an immutable bean which should be easily serializable.

boaks replied:
If the "pendingRequest" should not be "serialized", I think it's possible to exclude them.

May idea about this was very simple:
- the server already has a map with the <reg-id> as key. And the access to such a map is not well scaling.
- so just reuse that "already made access" by storing the stuff in the objects maintained by the first map.

sbernard31 replied:
> so just reuse that "already made access" by storing the stuff in the objects maintained by the first map.
Which map ? (could you give the name of attribute and class of map you are talking about)

boaks replied:
Sure, "ClientRegistryImpl.clientsByEp".

boaks replied:
OK, this is "endpoint name" to client, not "register id".
So staying with the "reg-id", the entry in the map should be removed on "cancelPendingRequests", (otherwise this may be something like a memory leak).
Alternatively, we use the endpoint name, copy the request list on cancel and iterate over the copy. I will make a purposal.

boaks replied:
PR #103 
Is a approach using concurrent collections instead of synchronization staying with the "reg-id" and is close to the original fix.

sbernard31 replied:
The `clientsByEp` is a map of the in memory `ClientRegistry` implementation. So you can not consider that this class will be always here. We have a [redis version](https://github.com/eclipse/leshan/blob/redis/leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java#L50) in development  where this map does not exists.
For me `ClientRegistry` should not handle pending request this is not its responsibility.

boaks replied:
Sure, the `ClientRegistry` should not handle pending request, but it's the central map for `Client`, regardless of its implementation.
The `CaliforniumLwM2mRequestSender.send` therefore has already a `Client` as parameter, so in my idea there is not to use a "second map/lookup" and instead use that Client. If that `Client` is extended by a List of `RequestCanceler` (see PR #102) it would be easy to cancel that list. But it was just an idea. Though this PR ist a Bugfix, I think it's better to find a solution then to discuss possibilities over and over. Therefore my proposal (see PR #103) is much closer to this solution than my "idea".

sbernard31 replied:
Totally agree, we should  keep focus on this bugfix PR. 
(About the use of Client object, I see what you mean, but this will not work with a not in memory implementation because you could have 2 `Client` instances which will represent the same Client/Registration. Each get on ClientRegistry will return a new instance)

boaks replied:
Each get on ClientRegistry will return a new instance

OK, I see.

boaks commented on leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java:
```diff
@@ -108,23 +117,102 @@ public T buildResponse(final Response coapResponse) {
+
+        @Override
+        public void onAcknowledgement() {
+            removePendingRequest(registrationId, coapRequest);
```
This will "fix/bind" the usagee of piggybacked reponses.
Though "separate responses" have also advantages, I'm not sure, if binding to hard to piggybacked is a good idea.

sbernard31 replied:
The main idea was to avoid the retransmission. That's why I firstly chose to remove pending request on acknowledgement, but maybe this is better to just wait the response.

boaks commented:
Is "cancel request on deregister" only wanted for californium?
If not, shouldn't we introduce a "cancel" into the core?
(OK, I already have a experimental extension for the client side. I don't want to have to many PRs open, but just for a "impression" I will create one and comment it with "don't merge".)

boaks commented on leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/RegistrationTest.java:
```diff
@@ -86,6 +90,38 @@ public void register_update_deregister() {
+        assertNull("No response expected", callback.getResponse());
+        assertNotNull("Exception expected", callback.getException());
+    }
+
```
May be add a test with multiple reads will show, if the "remove while iterate" is working.

boaks replied:
@Test
    public void deregister_cancel_multiple_pending_request() throws InterruptedException {
        // Check registration
        assertTrue(helper.server.getClientRegistry().allClients().isEmpty());

        helper.client.start();
        helper.waitForRegistration(1);

        assertEquals(1, helper.server.getClientRegistry().allClients().size());
        Client client = helper.server.getClientRegistry().get(ENDPOINT_IDENTIFIER);
        assertNotNull(client);
        assertArrayEquals(LinkObject.parse("</>;rt=\"oma.lwm2m\",</1/0>,</2>,</3/0>".getBytes()),
                client.getObjectLinks());

        // Stop client with out de-registration
        helper.client.stop(false);

        // Send multiple reads which should be retransmitted.
        List<Callback<ReadResponse>> callbacks = new ArrayList<Callback<ReadResponse>>();

        for (int index = 0; index < 4; ++index) {
            Callback<ReadResponse> callback = new Callback<ReadResponse>();
            helper.server.send(client, new ReadRequest(3, 0, 1), callback, callback);
            callbacks.add(callback);
        }

        // Restart client (de-registration/re-registration)
        helper.client.start();

        // Check the request was cancelled.
        int index = 0;
        for (Callback<ReadResponse> callback : callbacks) {
            boolean timedout = !callback.waitForResponse(1000);
            assertFalse("Response or Error expected, no timeout, call " + index, timedout);
            assertTrue("Response or Error expected, call " + index, callback.isCalled().get());
            assertNull("No response expected, call " + index, callback.getResponse());
            assertNotNull("Exception expected, call " + index, callback.getException());
            ++index;
        }
    }

Debugging this test shows:
- remove during iterate throws an exception
- that exception is "silently ignored"

sbernard31 replied:
At my side, this test works too. (and I can not find any silenty ignored exception, but maybe I don't find at the good place)
But in all case, I agree with you we must not remove pending iteration using foreach.

boaks replied:
I only analysed the deregister_cancel_multiple_pending_request() and that shows the exception.

But though you agree, I guess the fix will make this test workign as well.

sbernard31 commented:
> Would it be possible to put this change on an "experimental" branch?

This is already in a branch [cancel_request](https://github.com/eclipse/leshan/tree/cancel_request).

sbernard31 commented:
> Is "cancel request on deregister" only wanted for californium?
For now that's was the idea, as we don't have cancel on request.
> If not, shouldn't we introduce a "cancel" into the core?
We could but this is a more impacting work.
> (OK, I already have a experimental extension for the client side. I don't want to have to many PRs open, but just for a "impression" I will create one and comment it with "don't merge".)
Sounds good. We will be able to look at this.

boaks commented:
Sounds good. We will be able to look at this.

See PR #102. The principle is very primitive. So it should be also possible for the server side.
I'm not sure about the "synchronous" calls and how they should be handled.

sbernard31 commented:
I hope I fixed the concurrency issue and also added the Achim's test..

boaks commented on leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java:
```diff
@@ -137,34 +139,41 @@ public T buildResponse(final Response coapResponse) {
+    private String getKey(String registrationId, Request coapRequest) {
+        if (registrationId == null || coapRequest == null)
+            return null;
+        return registrationId + '#' + coapRequest.hashCode();
```
I think, this relys on a unique `hashCode()` ...

sbernard31 replied:
The java doc say :
```java
    /**
     * ....
     * As much as is reasonably practical, the hashCode method defined by
     * class {@code Object} does return distinct integers for distinct
     * objects. (This is typically implemented by converting the internal
     * address of the object into an integer, but this implementation
     * technique is not required by the
     * Java<font size="-2"><sup>TM</sup></font> programming language.)
```
But maybe this is not the best approach... I thought about the CoAP messageID  but not sure this is better?
The CoAP spec say that:
"The same Message ID MUST NOT be reused (in communicating with the same endpoint) within the EXCHANGE_LIFETIME"_

jvermillard replied:
Use  default hashcode or MID. I won't be :bike: shedding on this one :)

boaks replied:
Though `addPendingRequest` always creates a `CleanerMessageObserver`, it should not be too hard to move the "key" (reg-id + unique id) to that class and then use that key in add and remove.
So there is no need to depend on `hashCode` of a third party library (even if we know, it's not really third party :-).

sbernard31 commented:
I think all comments were taken into account.
So I rebased the code and cleaned the history to make this more readable for the next reviewers. (@msangoi ?)

msangoi commented on leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/RegistrationTest.java:
```diff
@@ -86,6 +91,79 @@ public void register_update_deregister() {
+    }
+
+    @Test
+    public void deregister_cancel_multiple_pending_request() throws InterruptedException {
```
With this test, the first unit test with a single request (deregister_cancel_pending_request) is probably not necessary anymore.

msangoi commented on leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java:
```diff
@@ -108,23 +119,109 @@ public T buildResponse(final Response coapResponse) {
     }
 
+    private String getFloorKey(String registrationId) {
+        if (registrationId == null)
```
This check is not needed. If this method returns `null`, a NPE will be raised later (`concurrentSkipListMap.subMap()`).
The null-check can be moved to the `cancelPendingRequests()` method.
Same thing for `getCeilingKey()`.

sbernard31 commented:
I added comment about keys generation and also changed the `null` test by a `Validate.notNull()`
This is now integrated in master (commit ead8c0323da676f8cf1b39937b73e83a78b05838, and 292dc374d5074b9a2e0eb1aff7e296ab4a5e2076)
Thx all for feedback.
